File: backend/app/schemas/training.py
from typing import Annotated, Literal, Optional

from pydantic import BaseModel, Field

from app.schemas._common import (
    DateLikeStr,
    IdStr,
    OptDateLikeStr,
    OptDigitStr,
    OptIdStr,
    OptShortStr,
    OptTextStr,
    ShortStr,
)


class ModuleConfig(BaseModel):
    """One block of `conference.sessionConfig` - shared shape for the
    Standard Test, Live Quiz and Survey modules of the session flow."""

    category: OptShortStr = None
    assessmentSuiteUid: OptIdStr = None
    questionCount: Optional[Annotated[int, Field(ge=0, le=10_000)]] = None
    startTime: OptDateLikeStr = None
    endTime: OptDateLikeStr = None
    checkIn: bool = False
    unlockCondition: OptShortStr = None


class AssessmentSuiteOut(BaseModel):
    """One row of `assessmentsuite`, as offered to the trainer's Category /
    Select Question Set pickers when building a session flow."""

    assessmentSuiteUid: str
    category: str
    name: str
    noOfQuestion: int


class QuestionOptionIn(BaseModel):
    id: IdStr
    text: Annotated[str, Field(max_length=1000)]


class QuestionCreate(BaseModel):
    question: Annotated[str, Field(min_length=1, max_length=2000)]
    questionType: ShortStr = "multiple_choice"
    options: Annotated[list[QuestionOptionIn], Field(max_length=20)] = []
    correctAnswer: OptIdStr = None
    points: Annotated[int, Field(ge=0, le=1000)] = 1
    timerSeconds: Optional[Annotated[int, Field(ge=0, le=86_400)]] = None
    explanation: OptTextStr = None


class QuestionOut(BaseModel):
    id: int
    question: str
    questionType: str
    options: list[QuestionOptionIn]
    correctAnswer: Optional[str] = None
    points: int
    timerSeconds: Optional[int] = None
    explanation: Optional[str] = None
    sortOrder: int


class AssessmentSuiteCreate(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=200)]
    description: OptTextStr = None
    category: Annotated[str, Field(min_length=1, max_length=120)]
    testTime: OptDigitStr = None
    type: ShortStr = "Quiz"


class AssessmentSuiteDetail(BaseModel):
    assessmentSuiteUid: str
    title: str
    description: Optional[str] = None
    category: str
    testTime: Optional[str] = None
    type: str
    noOfQuestion: int
    questions: list[QuestionOut] = []


class PendingSessionItem(BaseModel):
    conferenceUid: str
    title: str
    trainerName: Optional[str] = None
    conferenceDate: Optional[str] = None
    conferenceTime: Optional[str] = None
    status: str


class AttendanceConfig(BaseModel):
    checkInOpens: Optional[str] = None
    checkOutCloses: Optional[str] = None
    geoFencing: bool = False
    # Metres a trainee may be from the venue and still check in. Only used when
    # geoFencing is on and the venue has coordinates. Defaults to 100 (the
    # `conference.geoRadius` column default) when not given.
    geoRadius: Optional[int] = Field(default=None, ge=10, le=5000)


class SessionFlowConfig(BaseModel):
    attendance: Optional[AttendanceConfig] = None
    standardTest: Optional[ModuleConfig] = None
    liveQuiz: Optional[ModuleConfig] = None
    survey: Optional[ModuleConfig] = None


class TrainingCreate(BaseModel):
    zone: OptShortStr = None
    region: OptShortStr = None
    company: OptShortStr = None
    requestedBy: OptShortStr = None

    trainerEmployeeId: OptIdStr = None
    trainerName: OptShortStr = None

    state: OptShortStr = None
    district: OptShortStr = None
    venue: OptIdStr = None

    isResidential: bool = False
    conferenceDate: DateLikeStr
    conferenceTime: DateLikeStr
    # Only meaningful when isResidential is set - the program's last day.
    # No dedicated conference column for this yet, so it's folded into
    # sessionConfig (see create_training) rather than conferenceEndsOn,
    # which already means something else (the session's actual end
    # timestamp, set by /trainings/{uid}/end).
    trainingEndDate: OptDateLikeStr = None
    trainingHub: OptShortStr = None
    audience: OptShortStr = None
    sessionType: OptShortStr = None
    trainingType: OptShortStr = None
    batchSize: OptDigitStr = None

    sessionFlow: Optional[SessionFlowConfig] = None
    checklist: Optional[Annotated[list[ShortStr], Field(max_length=100)]] = None


class TrainingOut(BaseModel):
    conferenceUid: str
    conferenceStatus: str
    status: str


class AttendanceMarkRequest(BaseModel):
    status: Literal["Present", "Absent", "Joined", "Pending"]
    # The trainer must give a reason for a manual Present/Absent mark - it's
    # appended (with who + when) to `attendance.remarks` as an audit trail.
    reason: str = Field(min_length=1, max_length=500)


class ProctoringUnlockRequest(BaseModel):
    # The trainer must give a reason to clear a proctoring lockout - appended
    # (with who + when) to `attendance.theftRemarks` + `remarks`.
    reason: str = Field(min_length=1, max_length=500)


class AttendanceListItemOut(BaseModel):
    """Powers the trainer's Attendance List / Pending Attendance / Confirmed
    Attendance screens - one row per attendance record across every
    conference this trainer owns. `marked`/`Present` vs not is how the
    Pending/Confirmed screens split the same list client-side."""

    attendanceId: str
    region: Optional[str] = None
    product: Optional[str] = None
    session: Optional[str] = None
    audienceType: Optional[str] = None
    conferenceDate: Optional[str] = None
    trainerName: Optional[str] = None
    trainerHoId: Optional[str] = None
    participantHoId: Optional[str] = None
    participantName: str
    phone: Optional[str] = None
    state: Optional[str] = None
    location: Optional[str] = None
    reportingManagerOfPromoter: Optional[str] = None
    attendanceStatus: str
    checkIn: Optional[str] = None
    checkOut: Optional[str] = None
    postTestScore: Optional[str] = None
    postTestScoreSummary: Optional[str] = None
    sessionTypeMethod: Optional[str] = None
    conferenceId: Optional[str] = None
    lastUpdates: Optional[str] = None
    marked: bool
    # How many of this trainer's trainings the participant is on, across all
    # of them - same value repeats on each of that participant's rows.
    trainerTrainingsTotal: int = 0
    trainerTrainingsPresent: int = 0
    trainerTrainingsPending: int = 0


class TrainingAgendaItem(BaseModel):
    conferenceUid: str
    title: str
    trainerName: Optional[str] = None
    hoid: Optional[str] = None
    conferenceDate: Optional[str] = None
    conferenceTime: Optional[str] = None
    conferenceStatus: str
    approvalStatus: str
    location: Optional[str] = None
    batchSize: Optional[str] = None
    trainingType: Optional[str] = None
    state: Optional[str] = None
    district: Optional[str] = None
    trainingHub: Optional[str] = None
    venueName: Optional[str] = None
    updatedBy: Optional[str] = None
    updationOn: Optional[str] = None
    timestamp: Optional[str] = None
    traineeCount: int = 0


class TrainerAgendaResponse(BaseModel):
    """GET /admin/trainings - the session list plus the dashboard's summary
    stats, all computed server-side so the frontend just displays them
    rather than re-deriving them from the raw list itself. `totalTrainees`
    is a de-duplicated headcount (see the router for how it's built from
    `attendance`/`assessment_results`); `totalSessions`/`completed`/`pending`
    are counted straight off the filtered `conference` rows. `recentCompleted`
    is always the trainer's most recently completed sessions all-time,
    regardless of `start`/`end` - the Recent Sessions card ignores whatever
    date scope the rest of this response uses."""

    trainings: list[TrainingAgendaItem]
    totalTrainees: int
    totalSessions: int
    completed: int
    pending: int
    executedPercentage: int
    pendingPercentage: int
    recentCompleted: list[TrainingAgendaItem] = []


class AudienceBreakdown(BaseModel):
    total: int
    present: int
    absent: int = 0
    # Joined via QR / on the list but the trainer hasn't marked them Present
    # or Absent yet.
    notMarked: int = 0
    # On this trainer's roster vs. walked-in / joined by QR.
    assigned: int = 0
    unassigned: int = 0
    # Participants with no earlier Present attendance in any other session -
    # being trained for the first time.
    fresh: int = 0


class AssessmentSummary(BaseModel):
    pass_: int = Field(0, alias="pass")
    fail: int
    totalAttempts: int

    model_config = {"populate_by_name": True}


class TopPerformer(BaseModel):
    traineeUid: str
    name: str
    score: float = 0
    maxScore: float = 0
    percentage: float


class SessionHeroStat(BaseModel):
    """Per-module summary for the Session Heroes cards (Live Quiz / Test)."""

    moduleKey: str  # "LIVE_QUIZ" | "STANDARD_TEST"
    label: str
    participants: int
    averagePercent: float
    bestPercent: float
    topName: Optional[str] = None


class TraineeRow(BaseModel):
    traineeUid: str
    name: str
    employeeId: Optional[str] = None
    phone: Optional[int] = None
    # "ASSIGNED" if on this trainer's roster, else "NOT ALLOCATED" (walked
    # in / joined by QR).
    audienceType: str = "NOT ALLOCATED"
    # "Present" | "Pending" (joined, not checked in) | "Absent" | "Attempted"
    status: str
    markedOn: Optional[str] = None
    checkOutTime: Optional[str] = None
    score: Optional[str] = None
    profilePhoto: Optional[str] = None
    # On-device proctoring lockout (post-test). `isLocked` drives the red
    # "LOCKED" pill + the trainer's unlock control; `proctoringLogs` are the
    # already-formatted strike/unlock lines from `attendance.theftRemarks`.
    isLocked: bool = False
    proctoringStrikes: int = 0
    proctoringMaxStrikes: int = 3
    proctoringLogs: list[str] = []


class ExecutionFlowItem(BaseModel):
    moduleKey: str
    label: str
    status: str  # "Pending" | "Running" | "Completed"
    startedAt: Optional[str] = None
    endedAt: Optional[str] = None
    elapsedSeconds: Optional[int] = None
    # Planned duration of this module in minutes, from the start/end times the
    # trainer set on it when building the session flow. None when it has no
    # usable pair. Feeds the "Assigned" budget + "Total Time Used" gauge.
    assignedMinutes: Optional[int] = None
    # True when the trainer may start this module now: the session is
    # running, nothing else is live, this module hasn't run yet, and every
    # module before it in the flow has finished. Drives the per-row Start
    # button on the Session Dashboard's Execution Flow.
    canStart: bool = False
    # True when the trainer may re-run this finished module now: the session
    # is running and nothing else is currently live. Drives the per-row
    # Restart button.
    canRestart: bool = False


class AuditLogEntry(BaseModel):
    """One STARTED/STOPPED pair for a module - a module can run more than
    once (e.g. if `advance-module` cycles back around), so unlike
    `ExecutionFlowItem` this is per-run, not per-module."""

    moduleKey: str
    label: str
    runNumber: int
    startedAt: Optional[str] = None
    endedAt: Optional[str] = None
    elapsedSeconds: Optional[int] = None
    isRunning: bool = False
    startedBy: Optional[str] = None
    # Free-text detail for a non-module event (e.g. the reason a trainer gave
    # when overriding a late session start). None for ordinary module runs.
    note: Optional[str] = None


class LiveBroadcastRequest(BaseModel):
    questionId: int


class LiveStudioQuestionOut(BaseModel):
    """One question of the configured Live Quiz suite, as shown in the
    trainer's Live Studio card while LIVE_QUIZ is the active module."""

    id: int
    order: int
    text: str
    timerSeconds: int
    points: int
    responseCount: int
    isActive: bool


class LiveStudioOut(BaseModel):
    """Only populated on SessionDashboardOut when `activeModuleId == "LIVE_QUIZ"`
    - drives the Live Studio broadcast console. `state` mirrors
    `conference.liveQuizState`; `timerEndsAt` is epoch milliseconds."""

    suiteUid: str
    suiteTitle: str
    state: str
    activeQuestionId: Optional[int] = None
    timerEndsAt: Optional[int] = None
    # Server clock at response time (epoch ms). The client subtracts its own
    # Date.now() to get the clock offset, so the countdown is right even when
    # the device clock disagrees with the server's.
    serverNowMs: Optional[int] = None
    participants: int
    totalResponses: int
    questions: list[LiveStudioQuestionOut] = []


class SessionDashboardOut(BaseModel):
    conferenceUid: str
    title: str
    trainingType: Optional[str] = None
    conferenceDate: Optional[str] = None
    conferenceTime: Optional[str] = None
    trainerName: Optional[str] = None
    location: Optional[str] = None
    conferenceStatus: str
    approvalStatus: str
    activeModuleId: Optional[str] = None
    # Question count of the active module's suite - only for STANDARD_TEST /
    # LIVE_QUIZ; None for ATTENDANCE / SURVEY / no active module.
    activeModuleQuestionCount: Optional[int] = None

    actualStartedAt: Optional[str] = None
    actualEndedAt: Optional[str] = None
    runtimeSeconds: Optional[int] = None

    audience: AudienceBreakdown
    assessment: AssessmentSummary
    topPerformers: list[TopPerformer]
    trainees: list[TraineeRow]
    executionFlow: list[ExecutionFlowItem] = []
    auditLog: list[AuditLogEntry] = []
    sessionHeroes: list[SessionHeroStat] = []
    # Only present while LIVE_QUIZ is the active module - see _live_studio in
    # services/training_service.py.
    liveStudio: Optional[LiveStudioOut] = None


class SessionReportParticipant(BaseModel):
    userId: str
    name: str
    role: str = "Participant"
    checkIn: Optional[str] = None
    checkOut: Optional[str] = None
    score: Optional[str] = None


class SessionReportSummary(BaseModel):
    conferenceId: str
    sessionName: str
    date: Optional[str] = None
    state: Optional[str] = None
    schedule: Optional[str] = None
    duration: Optional[str] = None
    venueLink: Optional[str] = None


class SessionReportOut(BaseModel):
    """Session Report screen. `standardTest` / `liveQuiz` list only the
    trainees who actually completed that module for this conference (a
    Submitted assessment result for the module's suite)."""

    summary: SessionReportSummary
    standardTest: list[SessionReportParticipant] = []
    liveQuiz: list[SessionReportParticipant] = []

File: src/api/training.ts
import { SelectOption } from "@/components/ui/SearchableSelect";
import { apiRequest, apiUpload } from "./client";
// All types are preserved — they are used extensively across screens.

export type ModuleConfig = {
  category?: string;
  assessmentSuiteUid?: string;
  questionCount?: number;
  startTime?: string;
  endTime?: string;
  checkIn: boolean;
  unlockCondition?: string;
};

export type AssessmentSuiteOut = {
  assessmentSuiteUid: string;
  category: string;
  name: string;
  noOfQuestion: number;
};

export type AttendanceConfig = {
  checkInOpens?: string;
  checkOutCloses?: string;
  geoFencing: boolean;
  /** Metres from the venue a trainee may still check in (geoFencing on). */
  geoRadius?: number;
};

export type SessionFlowConfig = {
  attendance?: AttendanceConfig;
  standardTest?: ModuleConfig;
  liveQuiz?: ModuleConfig;
  survey?: ModuleConfig;
};

export type TrainingCreatePayload = {
  zone?: string;
  region?: string;
  company?: string;
  requestedBy?: string;

  trainerEmployeeId?: string;
  trainerName?: string;

  state?: string;
  district?: string;
  venue?: string;

  isResidential: boolean;
  conferenceDate: string;
  conferenceTime: string;
  trainingEndDate?: string;
  trainingHub?: string;
  audience?: string;
  sessionType?: string;
  trainingType?: string;
  batchSize?: string;

  sessionFlow?: SessionFlowConfig;
  checklist?: string[];
};

export type TrainingOut = {
  conferenceUid: string;
  conferenceStatus: string;
  status: string;
};

export type TrainingAgendaItem = {
  conferenceUid: string;
  title: string;
  trainerName: string | null;
  conferenceDate: string | null;
  conferenceTime: string | null;
  conferenceStatus: string;
  approvalStatus: string;
  location: string | null;
  batchSize: string | null;
  trainingType: string | null;
  state: string | null;
  trainingHub: string | null;
  traineeCount?: number;
  hoid: string | null;
  venueName: string | null;
  district: string | null;
  updatedBy: string | null;
  updationOn: string | null;
  timestamp: string | null;
};

export type TrainerAgendaResponse = {
  trainings: TrainingAgendaItem[];
  totalTrainees: number;
  totalSessions: number;
  completed: number;
  pending: number;
  executedPercentage: number;
  pendingPercentage: number;
  recentCompleted: TrainingAgendaItem[];
};

export type AudienceBreakdown = {
  total: number;
  present: number;
  absent: number;
  // Joined / on the list but not yet marked Present or Absent.
  notMarked: number;
  // On this trainer's roster vs. walked-in / joined by QR.
  assigned: number;
  unassigned: number;
  // First-timers - no earlier Present attendance in any other session.
  fresh: number;
};

export type AssessmentSummary = {
  pass: number;
  fail: number;
  totalAttempts: number;
};

export type TopPerformer = {
  traineeUid: string;
  name: string;
  score: number;
  maxScore: number;
  percentage: number;
};

export type SessionHeroStat = {
  moduleKey: string; // "LIVE_QUIZ" | "STANDARD_TEST"
  label: string;
  participants: number;
  averagePercent: number;
  bestPercent: number;
  topName: string | null;
};

export type TraineeRow = {
  traineeUid: string;
  name: string;
  employeeId: string | null;
  phone: number | null;
  audienceType: string;
  status: string; // "Present" | "Pending" | "Absent" | "Attempted"
  markedOn: string | null;
  checkOutTime: string | null;
  score: string | null;
  profilePhoto: string | null;
  // On-device proctoring lockout (post-test). Drives the red "LOCKED" pill +
  // the trainer's unlock control on the Participant Master List.
  isLocked: boolean;
  proctoringStrikes: number;
  proctoringMaxStrikes: number;
  proctoringLogs: string[];
};

export type ExecutionFlowStatus = "Pending" | "Running" | "Completed";

export type ExecutionFlowItem = {
  moduleKey: string;
  label: string;
  status: ExecutionFlowStatus;
  startedAt: string | null;
  endedAt: string | null;
  elapsedSeconds: number | null;
  assignedMinutes: number | null;
  // Backend says this module may be started now (session live, nothing else
  // running, every earlier module finished). Drives the per-row Start button.
  canStart: boolean;
  // Backend says this finished module may be re-run now (session live,
  // nothing else running). Drives the per-row Restart button.
  canRestart: boolean;
};

export type AuditLogEntry = {
  moduleKey: string;
  label: string;
  runNumber: number;
  startedAt: string | null;
  endedAt: string | null;
  elapsedSeconds: number | null;
  isRunning: boolean;
  startedBy: string | null;
  /** Free-text detail for a non-module event (e.g. a late-start override
   *  reason). null for ordinary module runs. */
  note: string | null;
};

export type LiveStudioQuestion = {
  id: number;
  order: number;
  text: string;
  timerSeconds: number;
  points: number;
  responseCount: number;
  isActive: boolean;
};

export type LiveStudio = {
  suiteUid: string;
  suiteTitle: string;
  state: "IDLE" | "WAITING" | "QUESTION_LIVE" | "LEADERBOARD" | "FINISHED" | string;
  activeQuestionId: number | null;
  timerEndsAt: number | null;
  // Server clock when this was sent (epoch ms) - for clock-skew-correct countdown.
  serverNowMs: number | null;
  participants: number;
  totalResponses: number;
  questions: LiveStudioQuestion[];
};

export type SessionDashboard = {
  conferenceUid: string;
  title: string;
  trainingType: string | null;
  conferenceDate: string | null;
  conferenceTime: string | null;
  trainerName: string | null;
  location: string | null;
  conferenceStatus: string;
  approvalStatus: string;
  activeModuleId: string | null;
  activeModuleQuestionCount: number | null;
  actualStartedAt: string | null;
  actualEndedAt: string | null;
  runtimeSeconds: number | null;
  audience: AudienceBreakdown;
  assessment: AssessmentSummary;
  topPerformers: TopPerformer[];
  trainees: TraineeRow[];
  executionFlow: ExecutionFlowItem[];
  auditLog: AuditLogEntry[];
  sessionHeroes: SessionHeroStat[];
  // Present only while LIVE_QUIZ is the active module.
  liveStudio: LiveStudio | null;
};

export function createTraining(token: string, payload: TrainingCreatePayload) {
  return apiRequest<TrainingOut>("/admin/trainings", {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
    body: JSON.stringify(payload),
  });
}

export function fetchTrainerAgenda(token: string, range?: { start?: string; end?: string; all?: boolean }) {
  const params = new URLSearchParams();
  if (range?.start) params.set("start", range.start);
  if (range?.end) params.set("end", range.end);
  if (range?.all) params.set("all_sessions", "true");
  const query = params.toString();
  return apiRequest<TrainerAgendaResponse>(`/admin/trainings${query ? `?${query}` : ""}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchSessionDashboard(token: string, conferenceUid: string) {
  return apiRequest<SessionDashboard>(`/admin/trainings/${encodeURIComponent(conferenceUid)}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export type SessionReportParticipant = {
  userId: string;
  name: string;
  role: string;
  checkIn: string | null;
  checkOut: string | null;
  score: string | null;
};

export type SessionReport = {
  summary: {
    conferenceId: string;
    sessionName: string;
    date: string | null;
    state: string | null;
    schedule: string | null;
    duration: string | null;
    venueLink: string | null;
  };
  standardTest: SessionReportParticipant[];
  liveQuiz: SessionReportParticipant[];
};

export function fetchSessionReport(token: string, conferenceUid: string) {
  return apiRequest<SessionReport>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/report`,
    { headers: { Authorization: `Bearer ${token}` } },
  );
}

export type StartTrainingLocation = {
  /** The trainer's own GPS at start - checked against the venue geofence. */
  latitude?: number;
  longitude?: number;
  /** Set only when the trainer corrects the venue location from the
   *  "you're not at the venue" prompt - persisted to the venue + conference. */
  venueLatitude?: number;
  venueLongitude?: number;
};

export function startTraining(
  token: string,
  conferenceUid: string,
  photo: { uri: string; name: string; type: string },
  location?: StartTrainingLocation,
  /** Reason for starting after the scheduled time - required by the backend
   *  (409 LATE_START) once the start time has passed, recorded on the
   *  activity log. */
  lateStartReason?: string,
) {
  const formData = new FormData();
  formData.append("photo", { uri: photo.uri, name: photo.name, type: photo.type } as unknown as Blob);
  const fields: [keyof StartTrainingLocation, number | undefined][] = [
    ["latitude", location?.latitude],
    ["longitude", location?.longitude],
    ["venueLatitude", location?.venueLatitude],
    ["venueLongitude", location?.venueLongitude],
  ];
  for (const [key, value] of fields) {
    if (value != null) formData.append(key, String(value));
  }
  if (lateStartReason?.trim()) formData.append("lateStartReason", lateStartReason.trim());
  return apiUpload<TrainingOut>(`/admin/trainings/${encodeURIComponent(conferenceUid)}/start`, formData, token);
}

export type UploadFile = { uri: string; name: string; type: string };

// Ending a session is a Security Check-Out: the trainer's face photo + the
// signed attendance sheet (PDF or image) are both required.
export function endTraining(
  token: string,
  conferenceUid: string,
  photo: UploadFile,
  attendanceSheet: UploadFile,
) {
  const formData = new FormData();
  formData.append("photo", { uri: photo.uri, name: photo.name, type: photo.type } as unknown as Blob);
  formData.append(
    "attendanceSheet",
    { uri: attendanceSheet.uri, name: attendanceSheet.name, type: attendanceSheet.type } as unknown as Blob,
  );
  return apiUpload<TrainingOut>(`/admin/trainings/${encodeURIComponent(conferenceUid)}/end`, formData, token);
}

// Manually opens one module. The trainer walks the flow forward one Start
// at a time - the backend rejects a start that's out of sequence or while
// another module is still live.
export function startModule(token: string, conferenceUid: string, moduleKey: string) {
  return apiRequest<TrainingOut>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/modules/${encodeURIComponent(moduleKey)}/start`,
    { method: "POST", headers: { Authorization: `Bearer ${token}` } },
  );
}

// Force-ends the currently live module without opening the next one. Never
// ends the session (only endTraining does that).
export function stopActiveModule(token: string, conferenceUid: string) {
  return apiRequest<TrainingOut>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/modules/stop-active`,
    { method: "POST", headers: { Authorization: `Bearer ${token}` } },
  );
}

// Re-runs a finished module. The backend rejects it while another module is
// still live.
export function restartModule(token: string, conferenceUid: string, moduleKey: string) {
  return apiRequest<TrainingOut>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/modules/${encodeURIComponent(moduleKey)}/restart`,
    { method: "POST", headers: { Authorization: `Bearer ${token}` } },
  );
}

// --- Live Quiz (FFF) broadcast console --------------------------------------

function liveQuizAction(token: string, conferenceUid: string, action: string, body?: unknown) {
  return apiRequest<SessionDashboard>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/live-quiz/${action}`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
      body: body === undefined ? undefined : JSON.stringify(body),
    },
  );
}

export function broadcastLiveQuestion(token: string, conferenceUid: string, questionId: number) {
  return liveQuizAction(token, conferenceUid, "broadcast", { questionId });
}

export function stopLiveTimer(token: string, conferenceUid: string) {
  return liveQuizAction(token, conferenceUid, "stop-timer");
}

export function showLiveLeaderboard(token: string, conferenceUid: string) {
  return liveQuizAction(token, conferenceUid, "leaderboard");
}

export function showLiveLobby(token: string, conferenceUid: string) {
  return liveQuizAction(token, conferenceUid, "lobby");
}

export function finishLiveQuiz(token: string, conferenceUid: string) {
  return liveQuizAction(token, conferenceUid, "finish");
}

export function markAttendance(
  token: string,
  conferenceUid: string,
  traineeUid: string,
  // Trainer confirming (Present) or overriding (Absent) a trainee's
  // attendance on the Participant Master List. "Absent" is a hard eject.
  status: "Present" | "Absent",
  reason: string,
) {
  return apiRequest<SessionDashboard>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/attendance/${encodeURIComponent(traineeUid)}`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
      body: JSON.stringify({ status, reason }),
    }
  );
}

export function resetAttendance(token: string, conferenceUid: string, traineeUid: string) {
  return apiRequest<SessionDashboard>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/attendance/${encodeURIComponent(traineeUid)}`,
    {
      method: "DELETE",
      headers: { Authorization: `Bearer ${token}` },
    }
  );
}

/** Clear a trainee's on-device proctoring lockout from the Participant Master
 *  List - the trainer must give a reason (audit-logged). */
export function unlockProctoring(
  token: string,
  conferenceUid: string,
  traineeUid: string,
  reason: string,
) {
  return apiRequest<SessionDashboard>(
    `/admin/trainings/${encodeURIComponent(conferenceUid)}/attendance/${encodeURIComponent(traineeUid)}/unlock`,
    {
      method: "POST",
      headers: { Authorization: `Bearer ${token}` },
      body: JSON.stringify({ reason }),
    }
  );
}

export function fetchAssessmentSuites(token: string) {
  return apiRequest<AssessmentSuiteOut[]>("/admin/assessment-suites", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchTrainerName(token: string, username: string) {
  return apiRequest<{ username: string; name: string }>(
    `/admin/trainers/${encodeURIComponent(username)}`,
    { headers: { Authorization: `Bearer ${token}` } }
  );
}

export function fetchTrainers(token: string, company?: string) {
  const query = company ? `?company=${encodeURIComponent(company)}` : "";
  return apiRequest<SelectOption[]>(`/admin/trainers${query}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchVenues(token: string, district?: string) {
  const query = district ? `?district=${encodeURIComponent(district)}` : "";
  return apiRequest<SelectOption[]>(`/admin/venues${query}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchChecklistItems(token: string) {
  return apiRequest<SelectOption[]>("/admin/checklist-items", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchTrainingHubs(token: string) {
  return apiRequest<SelectOption[]>("/admin/training-hubs", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchAudiences(token: string) {
  return apiRequest<SelectOption[]>("/admin/audiences", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchSessionTypes(token: string) {
  return apiRequest<SelectOption[]>("/admin/session-types", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchTrainingTypes(token: string) {
  return apiRequest<SelectOption[]>("/admin/training-types", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function fetchRequestedByOptions(token: string) {
  return apiRequest<SelectOption[]>("/admin/requested-by-options", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export type PendingSessionItem = {
  conferenceUid: string;
  title: string;
  trainerName: string | null;
  conferenceDate: string | null;
  conferenceTime: string | null;
  status: string;
};

export function fetchPendingTrainings(token: string) {
  return apiRequest<PendingSessionItem[]>("/admin/trainings/pending", {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function approveTraining(token: string, conferenceUid: string) {
  return apiRequest<TrainingOut>(`/admin/trainings/${encodeURIComponent(conferenceUid)}/approve`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function rejectTraining(token: string, conferenceUid: string) {
  return apiRequest<TrainingOut>(`/admin/trainings/${encodeURIComponent(conferenceUid)}/reject`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
  });
}

export type QuestionOption = { id: string; text: string };

export type QuestionOut = {
  id: number;
  question: string;
  questionType: string;
  options: QuestionOption[];
  correctAnswer: string | null;
  points: number;
  timerSeconds: number | null;
  explanation: string | null;
  sortOrder: number;
};

export type QuestionCreatePayload = {
  question: string;
  questionType?: string;
  options?: QuestionOption[];
  correctAnswer?: string;
  points?: number;
  timerSeconds?: number;
  explanation?: string;
};

export type AssessmentSuiteDetail = {
  assessmentSuiteUid: string;
  title: string;
  description: string | null;
  category: string;
  testTime: string | null;
  type: string;
  noOfQuestion: number;
  questions: QuestionOut[];
};

export type AssessmentSuiteCreatePayload = {
  title: string;
  description?: string;
  category: string;
  testTime?: string;
  type?: string;
};

export function createAssessmentSuite(token: string, payload: AssessmentSuiteCreatePayload) {
  return apiRequest<AssessmentSuiteDetail>("/admin/assessment-suites", {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
    body: JSON.stringify(payload),
  });
}

export function fetchAssessmentSuiteDetail(token: string, suiteUid: string) {
  return apiRequest<AssessmentSuiteDetail>(`/admin/assessment-suites/${encodeURIComponent(suiteUid)}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}

export function addAssessmentQuestion(token: string, suiteUid: string, payload: QuestionCreatePayload) {
  return apiRequest<AssessmentSuiteDetail>(`/admin/assessment-suites/${encodeURIComponent(suiteUid)}/questions`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
    body: JSON.stringify(payload),
  });
}

export function deleteAssessmentQuestion(token: string, suiteUid: string, questionId: number) {
  return apiRequest<AssessmentSuiteDetail>(
    `/admin/assessment-suites/${encodeURIComponent(suiteUid)}/questions/${questionId}`,
    {
      method: "DELETE",
      headers: { Authorization: `Bearer ${token}` },
    }
  );
}

export { ApiError } from "./client";
